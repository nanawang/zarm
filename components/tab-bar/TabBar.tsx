import React, { PureComponent, cloneElement } from 'react';
import classnames from 'classnames';
import { BaseTabBarProps } from './PropsType';
import TabBarItem from './TabBarItem';

export interface TabBarProps extends BaseTabBarProps {
  prefixCls?: string;
  className?: string;
}

class TabBar extends PureComponent<TabBarProps, any> {
  static Item: typeof TabBarItem;

  static defaultProps: TabBarProps = {
    prefixCls: 'za-tab-bar',
    visible: true,
  };

  onChildChange = (value) => {
    const { onChange } = this.props;
    if (typeof onChange === 'function') {
      onChange(value);
    }
  };

  getSelected = (index, itemKey) => {
    const { activeKey, defaultActiveKey } = this.props;
    if (!activeKey) {
      if (!defaultActiveKey && index === 0) {
        return true;
      }
      return defaultActiveKey === itemKey;
    }
    return activeKey === itemKey;
  };

  render() {
    const { visible, prefixCls, children, style } = this.props;
    const cls = classnames(prefixCls, {
      [`${prefixCls}--hidden`]: !visible,
    });
    const items = React.Children.map(children, (element: JSX.Element, index) => {
      return cloneElement(element, {
        key: index,
        disabled: element.props.disabled,
        onChange: () => this.onChildChange(element.props.itemKey),
        badge: element.props.badge,
        title: element.props.title,
        icon: element.props.icon,
        itemKey: element.props.itemKey || index,
        style: element.props.style,
        selected: this.getSelected(index, element.props.itemKey),
      });
    });
    return <div className={cls} style={style}>{items}</div>;
  }
}

export default TabBar;
