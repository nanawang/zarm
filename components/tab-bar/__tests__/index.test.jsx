import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import TabBar from '../index';

const activeIcon = (
  <div style={{
    width: '24px',
    height: '24px',
    background: 'url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat',
  }}
  />
);

const icon = (
  <div style={{
    width: '24px',
    height: '24px',
    background: 'url(//cdn-health.zhongan.com/zarm/home.svg) top left / 24px 24px no-repeat',
  }}
  />
);

describe('ActivityIndicator', () => {
  const onChange = jest.fn();
  it('renders correctly', () => {
    const wrapper = render(
      <TabBar onChange={onChange}>
        <TabBar.Item
          itemKey="home"
          title="主页"
          icon={icon}
          activeIcon={activeIcon}
        />
      </TabBar>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('with defaultActiveKey', () => {
    const wrapper = render(
      <TabBar defaultActiveKey="home">
        <TabBar.Item
          itemKey="home"
          title="主页"
          icon={icon}
          activeIcon={activeIcon}
        />
      </TabBar>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('with activeKey', () => {
    const wrapper = render(
      <TabBar activeKey="home">
        <TabBar.Item
          itemKey="home"
          title="主页"
          icon={icon}
          activeIcon={activeIcon}
        />
      </TabBar>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('onChange', () => {
    const wrapper = mount(
      <TabBar defaultActiveKey="home" onChange={onChange}>
        <TabBar.Item
          itemKey="home"
          title="主页"
          icon={icon}
          activeIcon={activeIcon}
        />
      </TabBar>,
    );

    wrapper.find('.za-tab-bar__item').first().simulate('click');
    expect(onChange).toBeCalledWith('home');
  });

  it('with badge onChange', () => {
    const wrapper = mount(
      <TabBar defaultActiveKey="home" onChange={onChange}>
        <TabBar.Item
          itemKey="badge"
          title="主页"
          icon={icon}
          activeIcon={activeIcon}
          badge={{ shape: 'circle', text: '3' }}
        />
      </TabBar>,
    );

    wrapper.find('.za-tab-bar__item').first().simulate('click');
    expect(onChange).toBeCalledWith('badge');
  });
});
