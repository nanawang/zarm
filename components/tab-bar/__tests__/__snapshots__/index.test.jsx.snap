// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityIndicator renders correctly 1`] = `
<div
  class="za-tab-bar"
>
  <div
    class="za-tab-bar__item za-tab-bar--active"
  >
    <div
      class="za-tab-bar__icon"
    >
      <div
        style="width:24px;height:24px;background:url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat"
      />
    </div>
    <div
      class="za-tab-bar__title"
    >
      主页
    </div>
  </div>
</div>
`;

exports[`ActivityIndicator with activeKey 1`] = `
<div
  class="za-tab-bar"
>
  <div
    class="za-tab-bar__item za-tab-bar--active"
  >
    <div
      class="za-tab-bar__icon"
    >
      <div
        style="width:24px;height:24px;background:url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat"
      />
    </div>
    <div
      class="za-tab-bar__title"
    >
      主页
    </div>
  </div>
</div>
`;

exports[`ActivityIndicator with defaultActiveKey 1`] = `
<div
  class="za-tab-bar"
>
  <div
    class="za-tab-bar__item za-tab-bar--active"
  >
    <div
      class="za-tab-bar__icon"
    >
      <div
        style="width:24px;height:24px;background:url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat"
      />
    </div>
    <div
      class="za-tab-bar__title"
    >
      主页
    </div>
  </div>
</div>
`;
